src/main/scala/cromiam/webservice/SubmissionSupport.scala

```diff
@@ -0,0 +1,139 @@
+import cromiam.auth.User.requireUser
+import akka.event.LoggingAdapter
+import spray.json.{JsObject, JsString, JsValue}
+import Collection.{CollectionLabelName, LabelsKey, validateLabels}
+
+import scala.concurrent.{ExecutionContextExecutor, Future}
+
+trait SubmissionSupport {
+  val cromwellClient: CromwellClient
+  val samClient: SamClient
+
+  val log: LoggingAdapter
+
+  implicit def executor: ExecutionContextExecutor
+  implicit val materializer: ActorMaterializer
+
+  // FIXME - getting pathPrefix to shrink this keeps hosing up, there's gotta be some way to do this
+  def submitRoute: Route = (path("api" / "workflows" / Segment) | path("api" / "workflows" / Segment / "batch")) { _ =>
+    post {
+      requireUser { user =>
+        log.info("Received submission request from user " + user.userId)
+        toStrictEntity(Timeout) {
+          extractSubmission(user) { submission =>
+            extractRequest { request =>
+              println("REQ IS " + request)
+              complete { forwardSubmissionToCromwell(user, submission.collection, request.withEntity(submission.entity)) }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private def forwardSubmissionToCromwell(user: User,
+                                          collection: Collection,
+                                          submissionRequest: HttpRequest): Future[HttpResponse] = {
+    log.info("Forwarding submission request for " + user.userId + " with collection " + collection.name + " to Cromwell")
+
+    def registerWithSam(collection: Collection): Future[Unit] = {
+      samClient.requestSubmission(user, collection) recoverWith {
+        case SamDenialException => Future.failed(SamDenialException)
+        case e => Future.failed(SamConnectionFailure("new workflow registration", e))
+      }
+    }
+
+    (for {
+      resp <- cromwellClient.forwardToCromwell(submissionRequest)
+      _ <- registerWithSam(collection)
+    } yield resp) recover {
+      case SamDenialException => SamDenialResponse
+    }
+  }
+}
+
+object SubmissionSupport {
+  def extractCollection(user: User): Directive1[Collection] = {
+    formField(CollectionNameKey.?).map(x => Collection(x.getOrElse(user.userId.value)))
+  }
+
+  def extractSubmission(user: User): Directive1[WorkflowSubmission] = {
+    (
+      extractCollection(user) &
+      formFields((
+        WorkflowSourceKey,
+        WorkflowTypeKey.?,
+        WorkflowTypeVersionKey.?,
+        WorkflowInputsKey.?,
+        WorkflowOptionsKey.?,
+        WorkflowDependenciesKey.as[ByteString].?)) &
+      extractLabels &
+      extractInputAux
+    ).as(WorkflowSubmission)
+  }
+
+  def extractLabels: Directive1[Option[Map[String, JsValue]]] = {
+    formField(LabelsKey.?) flatMap validateLabels
+  }
+
+  def extractInputAux: Directive1[Map[String, String]] = {
+    formFieldMap.map(_.filterKeys(_.startsWith(WorkflowInputsAuxPrefix)))
+  }
+
+  final case class WorkflowSubmission(collection: Collection,
```

Comment:
ToL is there such a type in the Cromwell-verse that could be re-used?

Reply:
Mostly. `collection` is  purely a CromIAM concept.

However there are a lot of places where this could be merged in w/ Cromwell code, but it was always going to require some Cromwell changes as well so wanted to save it for followup.

Side note: In general the explorations I did on these topics revealed that we were replicating the same sort of info a few places in Cromwell as well, which didn't improve my desire to go fix it right now